Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/SteamEngine.cs

```diff
@@ -0,0 +1,1037 @@
+                            SEList[i].InitFromMSTS();
+                            SEList[i].Initialize();
+                        }
+                    }
+                    break;
+
+            }
+        }
+
+        public void Copy(SteamEngines other)
+        {
+            SEList = new List<SteamEngine>();
+            foreach (SteamEngine se in other.SEList)
+            {
+                SteamEngine steamEngine = new SteamEngine(Locomotive);
+                steamEngine.Copy(se);
+
+                SEList.Add(steamEngine);
+            }
+        }
+
+        public void Initialize()
+        {
+            foreach (SteamEngine se in SEList)
+            {
+                se.Initialize();
+            }
+        }
+
+        public void InitializeMoving()
+        {
+            foreach (SteamEngine se in SEList)
+            {
+                se.InitializeMoving();
+            }
+        }
+
+        /// <summary>
+        /// Saves status of each auxiliary on the list
+        /// </summary>
+        /// <param name="outf"></param>
+        public void Save(BinaryWriter outf)
+        {
+            outf.Write(SEList.Count);
+            foreach (SteamEngine se in SEList)
+                se.Save(outf);
+        }
+
+        /// <summary>
+        /// Restores status of each auxiliary on the list
+        /// </summary>
+        /// <param name="inf"></param>
+        public void Restore(BinaryReader inf)
+        {
+            int count = inf.ReadInt32();
+            if (SEList.Count == 0)
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    SEList.Add(new SteamEngine(Locomotive));
+                    SEList[i].InitFromMSTS();
+                    SEList[i].Initialize();
+                }
+
+            }
+            foreach (SteamEngine se in SEList)
+                se.Restore(inf);
+        }
+
+
+
+        /// <summary>
+        /// Updates each auxiliary on the list
+        /// </summary>
+        /// <param name="elapsedClockSeconds">Time span within the simulation cycle</param>
+        public void Update(float elapsedClockSeconds)
+        {
+            foreach (SteamEngine de in SEList)
+            {
+                de.Update(elapsedClockSeconds);
+            }
+        }
+
+        public void HandleEvent(PowerSupplyEvent evt)
+        {
+            foreach (SteamEngine de in SEList)
+            {
+                de.HandleEvent(evt);
+            }
+        }
+
+        public void HandleEvent(PowerSupplyEvent evt, int id)
+        {
+            if (id >= 0 && id < SEList.Count)
+            {
+                SEList[id].HandleEvent(evt);
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return (IEnumerator)GetEnumerator();
+        }
+
+        public SteamEnum GetEnumerator()
+        {
+            return new SteamEnum(SEList.ToArray());
```

Comment:
This has terrible performance issues, please use the solution in LocomotiveAxles class. See #813 to see problems with enumerators